### DISCUSSION TITLE
chunks archived to wrong chunkservers

### DISCUSSION THREAD

**Discussion by solj** [moosefs/moosefs#388] [Q&A]
## Have you read through available documentation and open Github issues?

Yes

## Is this a BUG report, FEATURE request, or a QUESTION? Who is the indended audience?

BUG report

#### Your moosefs version and its origin (moosefs.com, packaged by distro, built from source, ...).

3.0.114 RPMs from the MooseFS yum repository.

#### Operating system (distribution) and kernel version.

```
# uname -vsrm
Linux 4.18.0-193.14.2.el8_2.x86_64 #1 SMP Sun Jul 26 03:54:29 UTC 2020 x86_64
# mfsmount -V
MFS version 3.0.114-1
using FUSE kernel interface version 7.26
# cat /etc/centos-release
CentOS Linux release 8.2.2004 (Core)
```

#### Hardware / network configuration, and underlying filesystems on master, chunkservers, and clients.

All master, metalogger, and chunkserver instances are running on separate physical hardware.

* 1 master
* 2 metaloggers
* 7 chunkservers
  * 2 chunkservers with NVMe drives
  * 3 chunkservers with 12 10TB drives
  * 2 chunkservers with 40 16TB drives

#### How much data is tracked by moosefs master (order of magnitude)?

 - All fs objects: 549374658
 - Total space: 1.6 PiB
 - Free space: 1.0 PiB
 - RAM used: 123 GiB
 - last metadata save duration: ~16.1m

## Describe the problem you observed.

Chunks which _should_ be *archived* to the 2 40x16TB servers are instead being stored on the 2xNVMe chunkservers which are members of the *create* tier.

#### Can you reproduce it? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Yes. You can see below that the NVMe chunkservers are assigned label *F*. The 12x10TB chunkservers are assigned label *M* and the 40x16TB chunkservers are assigned label *S*.

```
# mfscli -n -SCS
+-------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                     Chunk Servers                                                                     |
+----------------+------+----+--------+---------+------+-------------+----------------------------------------+-----------------------------------------+
|                |      |    |        |         |      |             |          'regular' hdd space           |      'marked for removal' hdd space     |
|    ip/host     | port | id | labels | version | load | maintenance +-----------+---------+---------+--------+--------+--------+------+-------+--------+
|                |      |    |        |         |      |             |   chunks  |   used  |  total  | % used | status | chunks | used | total | % used |
+----------------+------+----+--------+---------+------+-------------+-----------+---------+---------+--------+--------+--------+------+-------+--------+
| 10.247.245.136 | 9422 |  6 |    A,F | 3.0.114 |    0 |         off |   1358222 | 508 GiB |  56 TiB |  0.89% |      - |      0 |  0 B |   0 B |      - |
| 10.247.245.137 | 9422 |  1 |    A,M | 3.0.114 |   39 |         off |  14479935 |  26 TiB | 109 TiB | 23.91% |      - |      0 |  0 B |   0 B |      - |
| 10.247.245.138 | 9422 |  2 |    A,M | 3.0.114 |   42 |         off |  14524298 |  26 TiB | 109 TiB | 23.91% |      - |      0 |  0 B |   0 B |      - |
| 10.247.245.139 | 9422 |  3 |    A,M | 3.0.114 |   40 |         off |  14609064 |  26 TiB | 109 TiB | 23.91% |      - |      0 |  0 B |   0 B |      - |
| 10.247.245.140 | 9422 |  5 |    A,S | 3.0.114 |    2 |         off | 108923000 | 256 TiB | 582 TiB | 44.01% |      - |      0 |  0 B |   0 B |      - |
| 10.247.245.141 | 9422 |  7 |    A,F | 3.0.114 |    0 |         off |   1358068 | 508 GiB |  56 TiB |  0.89% |      - |      0 |  0 B |   0 B |      - |
| 10.247.245.142 | 9422 |  4 |    A,S | 3.0.114 |    2 |         off | 109980746 | 257 TiB | 582 TiB | 44.18% |      - |      0 |  0 B |   0 B |      - |
+----------------+------+----+--------+---------+------+-------------+-----------+---------+---------+--------+--------+--------+------+-------+--------+
```

Below you can see I have created a storage class *fasthome* which should be the following:

| label      | definition |
| ----------|--------------|
| create   | 2 copies on *F* |
| keep     | 2 copies on *F* and 1 on *M* |
| archive | 2 copies on *S* |

You can also see a file which is archived is stored on chunkservers with the *create* label.

```
# mfsgetsclass .
.: 'fasthome'
# mfsscadmin list -l | grep fasthome
fasthome : 2->3->2 ; admin_only: NO ; mode: STD ; create_labels: [F] , [F] ; keep_labels: [F] , [F] , [M] ; arch_labels: [S] , [S] ; arch_delay: 1d
# mfschkarchive bar
bar: all chunks are archived
# mfsfileinfo bar
bar:
        chunk 0: 00000000091945C3_00000001 / (id:152651203 ver:1)
                copy 1: 10.247.245.136:9422 (status:VALID)
                copy 2: 10.247.245.141:9422 (status:VALID)
```

**Comment by chogata** [comment]
How soon after archiving did you check the info? I guess pretty soon, like a few minutes after?

This is not an instant process. In your storage class, when the chunks acquire archive status, they start to be "wrong-labeled" and "overgoal". The next time the chunk loop reaches such a chunk, it will "do something to repair the status". I this case, the overgoal status will take precedence over the wrong-labeled status, so first thing to do will be attempt to delete the unnecessary copy. Then, when this status is fixed (it doesn't have to happen immediately, if your system is deleting a lot and has reached the limits of deletions, this operation may be delayed), the next time the loop sees the chunk, it will attempt to replicate the chunk to correct labels. Again: attempt; if your system is busy and replication limits are filled, this operation may also be delayed. Moving chunks to correct labels is not a high priority job, so depending on how busy your installation is, it may take a while.

The chunk you have here was stored at F,F,M and when it was archived, the system deleted the copy from M, leaving only the F,F copies, which happen to match your create labels. Eventually it will get around to replicating the remaining two copies to S,S.

**Comment by solj** [comment]
> How soon after archiving did you check the info? I guess pretty soon, like a few minutes after?

This was left over the weekend and nothing in the "Resources" tab shows "overgoal"

![image](https://user-images.githubusercontent.com/220792/95873034-1fa0ed00-0d35-11eb-962b-05b45dc2561b.png)

> 
> This is not an instant process. In your storage class, when the chunks acquire archive status, they start to be "wrong-labeled" and "overgoal". The next time the chunk loop reaches such a chunk, it will "do something to repair the status". I this case, the overgoal status will take precedence over the wrong-labeled status, so first thing to do will be attempt to delete the unnecessary copy. Then, when this status is fixed (it doesn't have to happen immediately, if your system is deleting a lot and has reached the limits of deletions, this operation may be delayed), the next time the loop sees the chunk, it will attempt to replicate the chunk to correct labels. Again: attempt; if your system is busy and replication limits are filled, this operation may also be delayed. Moving chunks to correct labels is not a high priority job, so depending on how busy your installation is, it may take a while.

Is there any way to force MooseFS to make the change instantaneously? I'm testing various things so that I can understand what happens before I make changes to a lot of data.

**Comment by chogata** [comment]
You can't make the change instantaneously. You are allowed, theoretically, to set archive bit to a million of files with one short command. If the system tried to replicate them to new, correct labels at that very moment, it would simply become totally unresponsive, being so busy copying millions of chunks. Any and all operations in MooseFS are done up to defined limits (either per period of time or as a number of concurrently happening operations), so that the system never becomes unresponsive. Those limits can be fine tuned to match your needs and strength of your hardware, but they will still exist.

Having said that, one file should not remain on wrong labeled servers for a whole weekend, unless something in your system is preventing the chunks from being moved. How busy is your system? Are you sure the file wasn't modified over the weekend (any modification, including it's metadata) - this would instantly take away the archive flag? Do you have any duplicated chunks in your system (happens when you move HDDs between chunk serves, which is NOT a recommended operation but can happen after hardware failure) ? Does the last chunk loop report any locked chunks (info in the tables on the bottom of your main CGI tab)? How long does your chunk loop take?

To give you an estimate: on a mostly idle system (some operations happen, but load on all chunk servers is less than 5) with 1.7 million chunks present, moving 100 files from 2 keep servers to 2 archive servers took less than 10 minutes, with some files being moved in under a minute, the rest being partially moved in under a minute (one copy). Basically, whichever files were not entirely "served" during current loop, were completed in the next loop.

**Comment by jkiebzak** [comment]
> How busy is your system? Are you sure the file wasn't modified over the weekend (any modification, including it's metadata) - this would instantly take away the archive flag? Do you have any duplicated chunks in your system (happens when you move HDDs between chunk serves, which is NOT a recommended operation but can happen after hardware failure) ? Does the last chunk loop report any locked chunks (info in the tables on the bottom of your main CGI tab)? How long does your chunk loop take?

These are all great questions, but what would be even better is having a way/tool/screen/syslog/debug/something to know *what* the system is working on and *why* chunks just sit there sometimes. I've also seen this type of scenario where chunks just sit there for weeks and I have **no** way of knowing exactly why (aka which of your questions apply).

Obviously, the master *knows why* it is not touching those chunks. So why can we query the master to get those details?

After that, a tool to allow you to manually prioritize an operation would be valuable. The "human" in the equation may know that its more important for particular files/chunks to have certain operations prioritized - such as migrating file[s] out of a particular class *in entirety*.

*Just wait longer* is not a very good answer.

**Comment by solj** [comment]
> You can't make the change instantaneously. You are allowed, theoretically, to set archive bit to a million of files with one short command. If the system tried to replicate them to new, correct labels at that very moment, it would simply become totally unresponsive, being so busy copying millions of chunks. Any and all operations in MooseFS are done up to defined limits (either per period of time or as a number of concurrently happening operations), so that the system never becomes unresponsive. Those limits can be fine tuned to match your needs and strength of your hardware, but they will still exist.
> 
> Having said that, one file should not remain on wrong labeled servers for a whole weekend, unless something in your system is preventing the chunks from being moved. How busy is your system?

> 

The system is not extremely busy. However, I now think I know why this is happening. I previously had all data stored on unlabeled chunkservers and recently moved everything to various storage classes. This operation, it seems, must have higher priority than the "archive" operation.

> Are you sure the file wasn't modified over the weekend (any modification, including it's metadata) - this would instantly take away the archive flag?

The file was definitely not modified over the weekend.

>  Do you have any duplicated chunks in your system (happens when you move HDDs between chunk serves, which is NOT a recommended operation but can happen after hardware failure) ? 

I believe we have duplicated chunks, due to the move mentioned above. Why is this not a recommended operation?

> Does the last chunk loop report any locked chunks (info in the tables on the bottom of your main CGI tab)? How long does your chunk loop take?

Yes, last chunk loop reports 6 locked chunks.

![image](https://user-images.githubusercontent.com/220792/96053127-8d880a00-0e44-11eb-9376-7a11aac0c126.png)

Looks like the chunk loop takes a little less than an hour.

> To give you an estimate: on a mostly idle system (some operations happen, but load on all chunk servers is less than 5) with 1.7 million chunks present, moving 100 files from 2 keep servers to 2 archive servers took less than 10 minutes, with some files being moved in under a minute, the rest being partially moved in under a minute (one copy). Basically, whichever files were not entirely "served" during current loop, were completed in the next loop.

This is what I was expecting and part of why we left it over the weekend.


On a related note, assuming everything above is correct, it seems like perhaps the way archived chunks are reported could be improved. The way it reads to me (as a user) is that the chunks are archived. However, it seems like there is a state not represented by any of the tools where a chunk is "to be archived" where the movement of data hasn't actually taken place. It would be nice to at least make that transparent to the user so we aren't left to guess why the data isn't in the proper location.

Thanks for the explanation!

**Comment by chogata** [comment]
Why moving disks is not a recommended operation - because it creates duplicates. Duplicates are an unnecessary drain on MooseFS resources, affecting it's performance. For example, when you (re)start a chunkserver with duplicates, it will not use the .chunkdb file mechanizm, meaning it will always make a full scan of all disks with chunks, which takes time and renders this one chunkserver basically useless until the scan is done. Also internal chunk server replication/balance operations are affected by duplicates (disks with duplicates are excluded from them), meaning a chunkserver may end up with unbalanced disk usage. And a chunkserver with unbalanced disk usage actively tries to restore the balance, using up system resources, but it will not succeed until duplicates are removed.

We introduced the possibility of moving disks between chunkservers for ONE reason only: to help in a situation when one or more chunkservers have a hardware failure and the user can't immediately find another hardware to replace them. If disks are unaffected (something else broke, like power supply, mainboard etc.) and contain data that is unavailable elsewhere in the system (AKA "last remaining copy of a chunk"), then the solution is to connect those drives to other, working chunkservers and wait until the system makes enough copies of those chunks, then disconnect and reformat them. 

Of course, we made the system so it manages even if those disks just stay there (the duplicates get deleted eventually, BTW, there is an option in master config to shorten that time), but, as I said, that affects the whole system.

TL;DR: moving disks between chunkservers was designed as an emergency solution ONLY and should not be used to do normal maintenance on the system.

> The system is not extremely busy. However, I now think I know why this is happening. I previously had all data stored on unlabeled chunkservers and recently moved everything to various storage classes. This operation, it seems, must have higher priority than the "archive" operation

Not really. As I explained above (a bit indirectly, maybe), "archiving" isn't really an operation in and of itself. Archiving simply means "switch on the archive flag". After the flag is switched on, if your chunks have different policies between keep and archive modes (different number of copies, different labels to keep the copies on), the system starts to treat them as chunks that need "cleaning up". In your case step one of cleanup was done (deleting the third copy), step two was supposed to be "move to correct labels". But, if you had a lot of chunks that needed to be moved to correct labels (because you just introduced labels into your previously unlabeled system), the order of those operations would be more or less random; the "archived" chunks don't get moved to correct labels any sooner than "unarchived" chunks that also happen to be on wrong-labeled servers, from the replication routine's point of view they are the same kind of operations.
